web/src/search.js

```diff
@@ -165,14 +175,57 @@ export function initialize() {
                 $search_query_box.trigger("blur");
                 update_buttons_with_focus(false);
             }
+
+            if (e.key === "Backspace" && $search_query_box.val() === "") {
+                // We just pressed Backspace and the box was empty. This
+                // dismisses the typeahead, which leaves us in a broken / ugly
+                // UI state due to the box shadow on the search query box. To
+                // prevent this, we trigger a lookup to bring the typeahead
+                // back.
+                open_typeahead();
+            }
         });
 
     // Some of these functions don't actually need to be exported,
     // but the code was moved here from elsewhere, and it would be
     // more work to re-order everything and make them private.
 
-    $search_query_box.on("focus", focus_search);
+    // We don't want to make this a focus handler because selecting the
+    // typehead seems to trigger this (and we don't want to open search
+    // when an option is selected and we're closing search).
+    // Instead we explicitly initiate search on click and on specific keyboard
+    // shortcuts.
+    // TODO(pre-merge): this still doesn't work great for keyboard navigation by tab,
+    // because the typeahead opens on focus but the search bar doesn't get
+    // fully expanded.
```

Comment:
I'm having some issues where if I  make this a focus handler, then I get a bug when I search for a new stream and click on it -- instead of closing the search bar and showing me the new stream header, it reopens an empty expanded search bar after going to the new narrow. I don't feel great about this solution though, and am seeking others' thoughts.